geo/src/algorithm/line_measures/metric_spaces/euclidean/distance.rs

```diff
@@ -0,0 +1,1087 @@
+            .min(
+                polygon
+                    .exterior()
+                    .lines()
+                    .map(|line| {
+                        ::geo_types::private_utils::line_segment_distance(
+                            point.0, line.start, line.end,
+                        )
+                    })
+                    .fold(Bounded::max_value(), |accum, val| accum.min(val)),
+            )
+    }
+}
+
+// ┌──────────────────────────┐
+// │ Implementations for Line │
+// └──────────────────────────┘
+
+symmetric_distance_impl!(CoordFloat, &Line<F>, Coord<F>);
+symmetric_distance_impl!(CoordFloat, &Line<F>, &Point<F>);
+
+impl<F: GeoFloat> Distance<F, &Line<F>, &Line<F>> for Euclidean {
+    fn distance(line_a: &Line<F>, line_b: &Line<F>) -> F {
+        if line_a.intersects(line_b) {
+            return F::zero();
+        }
+        // minimum of all Point-Line distances
+        Self::distance(&line_a.start_point(), line_b)
+            .min(Self::distance(&line_a.end_point(), line_b))
+            .min(Self::distance(&line_b.start_point(), line_a))
+            .min(Self::distance(&line_b.end_point(), line_a))
+    }
+}
+
+impl<F: GeoFloat> Distance<F, &Line<F>, &LineString<F>> for Euclidean {
+    fn distance(line: &Line<F>, line_string: &LineString<F>) -> F {
+        line_string
+            .lines()
+            .fold(Bounded::max_value(), |acc, segment| {
+                acc.min(Self::distance(line, &segment))
+            })
+    }
+}
+
+impl<F: GeoFloat> Distance<F, &Line<F>, &Polygon<F>> for Euclidean {
+    fn distance(line: &Line<F>, polygon: &Polygon<F>) -> F {
+        if line.intersects(polygon) {
+            return F::zero();
+        }
+
+        // REVIEW: This impl changed slightly.
+        std::iter::once(polygon.exterior())
+            .chain(polygon.interiors().iter())
+            .fold(Bounded::max_value(), |acc, line_string| {
+                acc.min(Self::distance(line, line_string))
+            })
+    }
+}
+
+// ┌────────────────────────────────┐
+// │ Implementations for LineString │
+// └────────────────────────────────┘
+
+symmetric_distance_impl!(CoordFloat, &LineString<F>, &Point<F>);
+symmetric_distance_impl!(GeoFloat, &LineString<F>, &Line<F>);
+
+impl<F: GeoFloat> Distance<F, &LineString<F>, &LineString<F>> for Euclidean {
+    fn distance(line_string_a: &LineString<F>, line_string_b: &LineString<F>) -> F {
+        if line_string_a.intersects(line_string_b) {
+            F::zero()
+        } else {
+            nearest_neighbour_distance(line_string_a, line_string_b)
+        }
+    }
+}
+
+impl<F: GeoFloat> Distance<F, &LineString<F>, &Polygon<F>> for Euclidean {
+    fn distance(line_string: &LineString<F>, polygon: &Polygon<F>) -> F {
+        if line_string.intersects(polygon) {
+            F::zero()
+        } else if !polygon.interiors().is_empty()
+            // FIXME: Explodes on empty line_string
```

Comment:
re: `FIXME` - this panic on the next line isn't new behavior in this PR, but I noticed it while moving the code over.

Note that currently `nearest_neighbor_distance` returns `Float::max_value` if the input geometries are empty.

So, since we're already returning `Float::max_value` for _some_ degenerate input, I'd argue it'd be preferable to **always** return `Float::max_value` for **any** degenerate input in this algorithm, rather than sometimes `panic`'ing and sometimes returning `Float::max_value`.

A larger breaking change could be to make distance return an `Option`.

❤️ empty geometries ❤️